gh-112622: Pass name to loop create_task method

ordinary-jamie opened the pull request:
Issue: https://github.com/python/cpython/issues/112622

The name parameter is not being passed to the event loop and is instead being set by the asyncio.create_task function. This means that third party implementations of the event loop will never receive the name parameter. This commit fixes this.

Note [Python 3.7 (now end-of-life) is the last version which does not expect a name parameter for `asyncio.create_task`](https://docs.python.org/3.7/library/asyncio-task.html#asyncio.create_task).


<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-112622 -->
* Issue: gh-112622
<!-- /gh-issue-number -->

ordinary-jamie committed:
Pass name to loop create_task method

The name parameter is not being passed to the event loop and is instead
being set by the asyncio.create_task function. This means that third
party implementations of the event loop will never receive the name
parameter. This commit fixes this. Note that python3.7 (eol) if the
last version of python that does not support the name parameter.

ordinary-jamie committed:
add news entry

gvanrossum commented on Lib/asyncio/tasks.py:
```diff
@@ -404,9 +404,9 @@ def create_task(coro, *, name=None, context=None):
-        task = loop.create_task(coro, context=context)
+        task = loop.create_task(coro, name=name, context=context)
 
     task.set_name(name)
```
If you're adding the name to the create_task() call, shouldn't you delete this line? It would seem to be redundant.

gvanrossum reviewed (commented):
Shouldn't you also do this for the C accelerator module `_asyncio`?

Also, create_task() exists mainly so that e.g. uvloop can override it. Are you sure that uvloop's create_task() takes a name keyword arg?

ordinary-jamie committed:
Pass name to loop create_task for TaskGroup

`TaskGroup.create_task` follows the same pattern as
`asyncio.create_task`, also including a redundant call to
`Task.set_name`. Note that `BaseEventLoop.create_task` uses the
`task_factory` if it is set, in which does not accept a `name` kwarg.
Hence, while we can remove the redundant call in `TaskGroup` we did not
remove it in the `asyncio.create_task`

gvanrossum commented on Lib/asyncio/taskgroups.py:
```diff
@@ -158,10 +158,10 @@ def create_task(self, coro, *, name=None, context=None):
+            task = self._loop.create_task(coro, name=name)
         else:
-            task = self._loop.create_task(coro, context=context)
-        task.set_name(name)
```
Hm. If you're keeping it in the other place it should be kept here too, shouldn't it? It's exactly the same pattern -- call `loop.create_task(coro, name=name [, context=context])`.

ordinary-jamie replied:
Sorry I think I'm confusing myself --

The call order for `create_task` from the `asyncio` module function would be (when the task factory is set):
```
asyncio.create_task(coro, name, context)
└── BaseEventLoop.create_task(coro, name, context)
    ├── BaseEventLoop._task_factory(coro, context)
    └── Task.set_name(name)
```

Similarly, for `TaskGroup`
```
TaskGroup.create_task(coro, name, context)
└── BaseEventLoop.create_task(coro, name, context)
    ├── BaseEventLoop._task_factory(coro, context)
    └── Task.set_name(name)
```

I think you were right from the outset -- we can remove the redundant call for both `asyncio.create_task` as well as `TaskGroup.create_task` since the [BaseEventLoop has the call when the task factory is set](https://github.com/python/cpython/blob/main/Lib/asyncio/base_events.py#L468), and [uses the `asyncio.Task` constructor with the name kwarg when the factory is not set](https://github.com/python/cpython/blob/main/Lib/asyncio/base_events.py#L458)

gvanrossum commented on Misc/NEWS.d/next/Library/2023-12-03-01-01-52.gh-issue-112622.1Z8cpx.rst:
```diff
@@ -0,0 +1,2 @@
+Ensure ``name`` parameter is passed to event loop in
+:func:`asyncio.create_task`
```
Trailing period.

gvanrossum reviewed (approved):
Yup, let's drop the redundant `set_name()` calls. I'll merge! Thanks for working through this carefully.
